compiler/z/codegen/ControlFlowEvaluator.cpp

```diff
@@ -338,21 +338,53 @@ xmaxxminHelper(TR::Node* node, TR::CodeGenerator* cg, TR::InstOpCode::Mnemonic c
    generateS390LabelInstruction(cg, TR::InstOpCode::label, node, cFlowRegionStart);
    cFlowRegionStart->setStartInternalControlFlow();
 
+   //Checking common Condition Code and branching to cFlowRegionEnd
    generateRREInstruction(cg, compareRROp, node, lhsReg, rhsReg);
    generateS390BranchInstruction(cg, TR::InstOpCode::BRC, branchCond, node, cFlowRegionEnd);
 
-   generateRREInstruction(cg, moveRROp, node, lhsReg, rhsReg);
-
-   TR::RegisterDependencyConditions* deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 2, cg);
+   //Label Symbol for handling possible NaN operand case for float/double
+   TR::LabelSymbol* nanLabel;
+   //For Double & Float operands only, branching to nanLabel for Condition Code 3 when operands are unordered
+   if (node->getOpCode().isDouble() || node->getOpCode().isFloat())
+      {
+      nanLabel = generateLabelSymbol(cg);
+      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_CC3, node, nanLabel);
+      }
 
+   //Move resulting operand to lhsReg as fallthrough for alternate Condition Code
+   generateRREInstruction(cg, moveRROp, node, lhsReg, rhsReg);
+   TR::RegisterDependencyConditions* deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 3, cg);
    deps->addPostConditionIfNotAlreadyInserted(lhsReg, TR::RealRegister::AssignAny);
    deps->addPostConditionIfNotAlreadyInserted(rhsReg, TR::RealRegister::AssignAny);
+   //Branching to cFlowRegionEnd to jump over NaN case
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+
+   //Handle NaN case
+   generateS390LabelInstruction(cg, TR::InstOpCode::label, node, nanLabel);
+   //Allocate GPR for holding immediate NaN value for float/double
+   TR::Register * nanReg = cg->allocateRegister(TR_GPR);
+   //Handle NaN case for doubles
+   if (node->getOpCode().isDouble())
+      {
+      //Load immediate DOUBLE_NAN value into GPR, then move it to result FPR
+      genLoadLongConstant(cg, node, DOUBLE_NAN, nanReg, NULL, deps, NULL);
+      generateRRInstruction(cg, TR::InstOpCode::LDGR, node, lhsReg, nanReg);
+      }
+   else
```

Comment:
If node is not double than, it can be int, long, short, byte and float. You can not assume it will be Float only.